### PR TITLE
Issue #2840: Implement Javadoc tag parsing to accurately parse Javadoc tags that span multiple lines

### PR DESCRIPTION
**Pull request by nanaze**
Issue #2840: Extract implementations of the block and inline parsers and fix the inline tag parser to properly parse inline tags that span multiple lines. Put tests on both

### CODE REVIEW

**Comment by nanaze** [comment]
This is a recreation of the PR in https://github.com/checkstyle/checkstyle/pull/4318 

Regarding comments in https://github.com/checkstyle/checkstyle/pull/4318#issuecomment-298790804 : 

I've put in a dependency on AutoValue (https://github.com/google/auto/). This is a small annotation processor library used heavily at Google for quickly making "value classes". In this PR, it's being used as a small intermediary value in TagUtils (see above). Given your existing use of Guava, I thought this might be fine, but if you're not okay with that dependency, that's fine -- I can rewrite the utility class manually. Please let me know.

I am aware of your work to eventually parse the Javadoc as part of the AST, but did this work knowing it would be eventually superceeded as this misbehavior is a major pain point for us (Java authors at Google) -- our autoformatter will auto-wrap @link tags over lines, which cause checkstyle to miss them as usages and then fail lint (stating that an import is unused).

I will also take a look at your checks above. Thank you for your help.

**Comment by romani** [comment]
New dependency is not ok, you need to show very good reason to let me change my mind. But you have very little chance as your code will not live long.

All CI should be green.

You need to pass code review process.

Be ready to do regression testing of your change of big oprnsource projects.

**Comment by nanaze** [comment]
I've found your docs on contributing and now understand your coverage/testing setup.

http://checkstyle.sourceforge.net/contributing.html

 `mvn verify` now works and all checks pass.

Branch is updated. 

> New dependency is not ok, you need to show very good reason to let me change my mind. But you have very little chance as your code will not live long.

Understood. AutoValue has been dropped and I've rewritten the value class manually.

> All CI should be green.

CI checks are currently pending. I will investigate if any fail.

> You need to pass code review process.
> Be ready to do regression testing of your change of big oprnsource projects.

Please let me know if there are any changes you need on my end.

**Comment by romani** [comment]
@nanaze , please squash commits in one, and we start review.

Please start to read about

**Comment by nanaze** [comment]
> @nanaze , please squash commits in one, and we start review.

Done.

> Please run regression diff testing of check behavior before and after fix. Share report with us for verification. 

Ran `~/contribution/checkstyle-tester$  groovy diff.groovy --localGitRepo /home/nanaze/checkstyle --baseBranch master --patchBranch inline-tags --config my_check.xml --listOfProjects projects-to-test-on.properties`

Posted resulting diff report here:
http://pages.cs.wisc.edu/~naze/checkstyle/reports/diff/

**Comment by romani** [comment]
@nanaze , sorry for delay in review, I feel unwell, I will give feedback till Monday.

Please extend a bit amount of projects in diff testing to see a difference. You can add any another project that has javadoc with cases we are try to resolve.

**Comment by rnveach** [comment]
@nanaze You will also need to rebase and fix conflicts in `JavadocUtils`. We made changes in there in preparation for a new check in sevntu.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/InlineTagUtils.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    /**
+     * @param source Source string.
+     * @param index An index into the string.
+     * @return A position in the source representing what line and column that index appears on.
+     */
+    public static LineColumn getLineColumnOfIndex(String source, int index) {
+        if (index < 0) {
+            throw new AssertionError("index must be positive");
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/InlineTagUtils.java]: IllegalArgumentException should be used. 
in following case too.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/InlineTagUtils.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+    /** Non instantiable. */
+    private InlineTagUtils() { }
+
+    /**
+     * Extract inline Javadoc tags from the given comment.
+     * @param lines The Javadoc comment (as lines).
+     * @return The extracted inline Javadoc tags.
+     */
+    public static List<TagInfo> extractInlineTags(String... lines) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/InlineTagUtils.java]: All other methods should be private.
Coverage should be archived by test cases for `extractInlineTags`

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockTagUtils.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.utils;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockTagUtils.java]: please move all new Utils classes to 
`package com.puppycrawl.tools.checkstyle.checks.javadoc.utils;`
I do not want to give users a signal that such implementations will be first class citizens.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtils.java [hunk 4]
```diff
@@ -140,100 +123,43 @@ private JavadocUtils() {
...
-        if (lineNumber == 0) {
-            blockTagPattern = BLOCK_TAG_PATTERN_FIRST_LINE;
-        }
-        else {
-            blockTagPattern = BLOCK_TAG_PATTERN;
-        }
-        return blockTagPattern;
-    }
+        final boolean getBlockTags = tagType == JavadocTagType.ALL
+            || tagType == JavadocTagType.BLOCK;
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtils.java]: please intend this like below(+4 or more from "=") to ease reading:
```
        final boolean getBlockTags = tagType == JavadocTagType.ALL
                                         || tagType == JavadocTagType.BLOCK;
```

**Comment by romani** [comment]
@nanaze , there will be release on this Sunday. It will be the last release of 7.X  , next release will be 8.X and we start  introducing breaking compatibility changes.  So if you use checkstyle by some plugins you will not be able to use your changes soon.

My recommendation is to provide all required updates and report before weekend.

**Comment by romani** [comment]
@nanaze , if no reply till 1 July - I will close this PR.

**Comment by romani** [comment]
@nanaze , ping, last warning before I close PR.

**Comment by nanaze** [comment]
Hello,

I am leaving for paternity leave in a few days and will not have a chance to update this PR. If you wish, you are welcome to make any edits you wish and merge.

Thanks,

Nathan

**Comment by romani** [comment]
thanks for reply.
I am closing PR for now to ease management of PRs, as you got a time to finish , you can reopen PR and continue.

**Pull request closed without merging**